lib/web_ui/lib/src/engine/surface/scene_builder.dart

```diff
@@ -97,6 +100,16 @@ class SurfaceSceneBuilder implements ui.SceneBuilder {
     if (matrix4.length != 16) {
       throw ArgumentError('"matrix4" must have 16 entries.');
     }
+    if (_surfaceStack.length == 1) {
+      // Top level transform contains view configuration to scale
+      // scene to devicepixelratio. Use identity instead since CSS uses
+      // logical device pixels.
+      if (!ui.debugEmulateFlutterTesterEnvironment) {
+        assert(matrix4[0] == window.devicePixelRatio &&
+           matrix4[5] == window.devicePixelRatio);
+      }
+      matrix4 = Matrix4.identity().storage;
```

Comment:
Actually, now that I think about it, the correct fix would be to inject a transform that scales by 1/DPR. This way we don't have to rely on a particular framework behavior. To avoid extra div nodes we can do something like:

```
if (the scene contains a root transform) {
  reuse the root transform by scale it by 1/DPR.
} else {
  apply 1/DPR to the `<flt-scene>` element.
}
```

In most cases we'll hit the `if` branch, not the `else` branch. Also in most cases the root transform will be scale by DPR. So the resulting root transform will automatically become identity and we won't apply any transform at all.

Reply:
The framework itself injects pushTransform using ViewConfiguration. Ideally this code would not live in framework but in engine. See view.dart:39. Or instead of calling pushTransform , framework would call SceneBuilder.setViewConfiguration in engine.

Reply:
Right, but you can use `dart:ui` without the Flutter framework where you can paint a scene without a root transform. This typically happens in tests and benchmarks. I don't expect there to be many apps that don't use the framework.